src/window.rs

```diff
@@ -0,0 +1,149 @@
+                } else {
+                    height_scale
+                }
+            };
+            let height = (scale * image_height as f32) as u32;
+            let width = (scale * image_width as f32) as u32;
+            let output_image = image::imageops::resize(image, width, height, image::FilterType::Triangle);
+            (width, height, output_image)
+    };
+
+    const CHANNEL_COUNT: u32 = 4;
+    let pitch  = output_image_width * CHANNEL_COUNT;
+    let mut img_raw = output_image.into_raw();
+    let surface_img = Surface::from_data(
+        &mut img_raw, 
+        output_image_width, 
+        output_image_height,
+        pitch, 
+        PixelFormatEnum::ABGR8888  // this format is necessary because sdl2 expects bits from highest to lowest
+    ).expect("couldn't converted image to surface");
+
+    let sdl = sdl2::init().expect("couldn't create sdl2 context");
+    let video_subsystem = sdl.video().expect("couldn't create video subsystem");
+    let mut window = video_subsystem
+        .window(title, window_width, window_height)
+        .position_centered()
+        .resizable()
+        .build()
+        .expect("window couldn't be created");
+    window.set_minimum_size(300, 300).expect("invalid minimum size for window");
+
+    let mut canvas = window.into_canvas().build().expect("Couldn't create CanvasBuilder");
+    let texture_creator = canvas.texture_creator();
+
+    let mut texture = texture_creator.create_texture_from_surface(surface_img)
+        .expect("couldn't create texture from surface");
+
+    // calculates new location for surface from window origin so that
+    // the image is centered in the window
+    let center_x = ((window_width - output_image_width) as f32 / 2.0_f32) as i32;
+    let center_y = ((window_height - output_image_height) as f32 / 2.0_f32) as i32; 
+    
+    // makes background white
+    canvas.set_draw_color(Color::RGB(255, 255, 255));
+    canvas.clear();
+
+    // displays image in the window
+    canvas.copy(
+        &texture, 
+        None, 
+        Rect::new(center_x, center_y, output_image_width, output_image_height))
+        .unwrap();
+    canvas.present();
+
+    // create and start events loop to keep window open until Esc 
+    let mut event_pump = sdl.event_pump().unwrap();
+    event_pump.enable_event(sdl2::event::EventType::Window);
+    'running: loop {
+        for event in event_pump.poll_iter() {
+            match event {
+                Event::Quit {..} |
+                    Event::KeyDown {keycode: Some(Keycode::Escape), .. } => {
+                        break 'running
+                    },
+                Event::Window { win_event: WindowEvent::Resized(x, y), .. } => {
+                    
+                    let x = x as u32;
+                    let y = y as u32;
+                    // resize image if necessary to fit into the window
+                    let (output_image_width, output_image_height, output_image) = 
```

Comment:
This looks suspiciously similar to the scaling code above - could you extract the scaling code into a function private function and use it in both places?

Reply:
Would it be fine to make it a nested function?

Reply:
Sure